src/webgpu/api/validation/validation_test.ts

```diff
@@ -1,17 +1,103 @@
 import { GPUTest } from '../../gpu_test.js';
+import {
+  DevicePool,
+  DeviceProvider,
+  TestOOMedShouldAttemptGC,
+  UncanonicalizedDeviceDescriptor,
+} from '../../util/device_pool.js';
 
 import { CommandBufferMaker, EncoderType } from './util/command_buffer_maker.js';
 
 const kResourceStateValues = ['valid', 'invalid', 'destroyed'] as const;
 export type ResourceState = typeof kResourceStateValues[number];
 export const kResourceStates: readonly ResourceState[] = kResourceStateValues;
 
+const mismatchedDevicePool = new DevicePool();
+
 /**
  * Base fixture for WebGPU validation tests.
  */
 export class ValidationTest extends GPUTest {
+  // Device mismatched validation tests require another GPUDevice different with the dafault
+  // GPUDevice of GPUTest, which is only used to create device mismatched objects.
+  private mismatchedProvider: DeviceProvider | undefined;
+  private mismatchedAcquiredDevice: GPUDevice | undefined;
+
+  /** GPUDevice for creating mismatched objects required by device mismatched validation tests. */
+  get mismatchedDevice(): GPUDevice {
+    assert(
+      this.mismatchedProvider !== undefined,
+      'No provider available right now; did you "await" selectMismatchedDeviceOrSkipTestCase?'
+    );
+    if (!this.mismatchedAcquiredDevice) {
+      this.mismatchedAcquiredDevice = this.mismatchedProvider.acquire();
+    }
+    return this.mismatchedAcquiredDevice;
+  }
+
+  /**
+   * Create other device different with current test device, which could be got by `.mismatchedDevice`.
+   * A `descriptor` may be undefined, which returns a `default` mismatched device.
+   * If the request descriptor or feature name can't be supported, throws an exception to skip the entire test case.
+   */
+  async selectMismatchedDeviceOrSkipTestCase(
+    descriptor:
+      | UncanonicalizedDeviceDescriptor
+      | GPUFeatureName
+      | undefined
+      | Array<GPUFeatureName | undefined>
+  ): Promise<void> {
+    if (this.mismatchedProvider) {
+      const oldProvider = this.mismatchedProvider;
+      this.mismatchedProvider = undefined;
+      await mismatchedDevicePool.release(oldProvider);
+      this.mismatchedAcquiredDevice = undefined;
+    }
+
+    if (descriptor === undefined) {
+      this.mismatchedProvider = await mismatchedDevicePool.reserve();
+    } else {
+      if (typeof descriptor === 'string') {
+        descriptor = { requiredFeatures: [descriptor] };
+      } else if (descriptor instanceof Array) {
+        descriptor = {
+          requiredFeatures: descriptor.filter(f => f !== undefined) as GPUFeatureName[],
+        };
+      }
+      this.mismatchedProvider = await mismatchedDevicePool.reserve(descriptor);
+    }
+
+    this.mismatchedAcquiredDevice = this.mismatchedProvider.acquire();
+  }
+
+  protected async finalize(): Promise<void> {
+    await super.finalize();
+
+    if (this.mismatchedProvider) {
```

Comment:
I'd like to make this all a helper as well since it's duplicated, but it's kind of complex to make sure the behavior doesn't change. OK to leave this as is for now, maybe I'll look at it after this lands.
```suggestion
    if (this.mismatchedProvider) {
      // TODO(kainino0x): Deduplicate this with code in GPUTest.finalize
```